drone_flightplan/terrain_following_waylines.py

```diff
@@ -0,0 +1,277 @@
+    tp : list
+        A list of actual waypoints (dicts from GeoJSON)
+
+    threshold : float
+        The allowable deviation from a consistent AGL in m
+
+    writer : csv.writer object
+        If None, does nothing. If a csv.writer object, creates a CSV file
+        that provides detailed information about the waypoints being checked
+        for deviation from AGL along the slope of each segment.
+
+    Returns:
+    --------
+    new_keeperpoints : list
+        A list of integer waypoint indexes
+    """
+    # Create 2-point segments between existing keeper waypoints; each segment
+    # will be examined to see if another one in the middle is needed
+    currentpoint = kp[0]
+    segments = []
+    for endpoint in kp[1:]:
+        segment = (tp[currentpoint - kp[0]], tp[endpoint - kp[0]])
+        segments.append(segment)
+        currentpoint = endpoint
+
+    new_keeperpoints = []
+
+    # Run through all the segments between keeperpoints and inject (from
+    # the entire list of original waypoints) those needed to keep AGL
+    # deviation below the threshold
+    for segment in segments:
+        fp = segment[0]['geometry']
+        lp = segment[1]['geometry']
+        run = distance(fp, lp)
+        rise = lp.z - fp.z
+        slope = 0
+        # If run is zero will get divide by zero error, check first
+        if run:
+            slope = rise / run
+        max_agl_difference = 0
+        max_agl_difference_point = 0
+        injection_point = None
+        points_to_traverse = segment[1]['index'] - segment[0]['index']
+
+        pointsinfo = []
+        pointsinfo.append([segment[0]['index'],fp,fp.z,fp.z,0,0,'first'])
+
+        for i in range(1, points_to_traverse):
+            pt = tp[i]['geometry']
+            z = round(pt.z, 2)
+            ptrun = round(distance(fp, pt),2)
+            expected_z = (round(fp.z + (ptrun * slope),2))
+            agl_difference = abs(round(z - expected_z,2))
+            if (agl_difference > max_agl_difference and
+                agl_difference > threshold):
+                max_agl_difference = agl_difference
+                max_agl_difference_point = tp[i]['index']
+                injection_point = i
+            pointsinfo.append([tp[i]['index'], tp[i]['geometry'],expected_z,
+                               z, ptrun, agl_difference])
+        pointsinfo.append([segment[1]['index'],lp,lp.z,lp.z,rise,run,'last'])
+        if injection_point:
+            new_segment = [segment[0], tp[injection_point], segment[1]]
+            for new_point in new_segment:
+                new_keeperpoints.append(new_point['index'])
+            pointsinfo[injection_point].append('injection point')
+        else:
+            for point in segment:
+                new_keeperpoints.append(point['index'])
+        if writer:
+            writer.writerows(pointsinfo)
+
+    return new_keeperpoints
+            
+if __name__ == "__main__":
```

Comment:
I quite like wrapping the logic in another function, like `convert_waypoints_to_waylines`, taking the arguments from argparse in __main__ as params.

That way you can either execute the file as intended here, or call the function from code somewhere

Reply:
Done, all actual logic is in a waypoints2waylines function, which accepts and returns GeoJSON-formatted dicts. 

Writing to the filesystem still done from __main__, as I assume that someone calling the logic from another module might prefer to deal with in-memory objects rather than files on the local filesystem!